packages/jazz-tools/src/tools/coValues/request.ts

```diff
@@ -0,0 +1,617 @@
+import {
+  CoValueCore,
+  CojsonInternalTypes,
+  CryptoProvider,
+  RawAccount,
+  cojsonInternals,
+} from "cojson";
+import z from "zod/v4";
+import {
+  AnyCoMapSchema,
+  AnyCoSchema,
+  CoMapInitZod,
+  CoMapSchema,
+  CoValue,
+  CoValueClass,
+  Group,
+  Loaded,
+  ResolveQuery,
+  ResolveQueryStrict,
+  Simplify,
+  anySchemaToCoSchema,
+  coMapDefiner,
+  exportCoValue,
+  importContentPieces,
+  loadCoValue,
+} from "../internal.js";
+import { Account } from "./account.js";
+
+type MessageShape = Record<string, z.core.$ZodType | AnyCoSchema>;
+
+type RequestSchemaDefinition<
+  S extends MessageShape,
+  R extends ResolveQuery<CoMapSchema<S>> = true,
+> =
+  | S
+  | {
+      schema: S;
+      resolve?: R;
+    };
+
+/**
+ * Configuration options for defining HTTP request/response schemas in Jazz.
+ *
+ * This interface defines the structure for creating typed HTTP routes with
+ * request and response validation using CoMap schemas.
+ *
+ * @template RequestShape - The shape of the request message schema (must extend MessageShape)
+ * @template RequestResolve - The resolve query type for the request CoMap schema
+ * @template ResponseShape - The shape of the response message schema (must extend MessageShape)
+ * @template ResponseResolve - The resolve query type for the response CoMap schema
+ */
+interface RequestOptions<
+  RequestShape extends MessageShape,
+  RequestResolve extends ResolveQuery<CoMapSchema<RequestShape>>,
+  ResponseShape extends MessageShape,
+  ResponseResolve extends ResolveQuery<CoMapSchema<ResponseShape>>,
+> {
+  /**
+   * The URL endpoint for the HTTP route.
+   * This is used by the client to send requests to the server.
+   */
+  url: string;
+
+  /**
+   * The id of the worker Account or Group.
+   */
+  workerId: string;
+
+  /**
+   * Schema definition for the request payload.
+   * Can be either a direct schema object or an object with schema and optional resolve properties.
+   * The schema defines the structure and validation rules for incoming requests.
+   */
+  request: RequestSchemaDefinition<
+    RequestShape,
+    ResolveQueryStrict<CoMapSchema<RequestShape>, RequestResolve>
+  >;
+
+  /**
+   * Schema definition for the response payload.
+   * Can be either a direct schema object or an object with schema and optional resolve properties.
+   * The schema defines the structure and validation rules for outgoing responses.
+   */
+  response: RequestSchemaDefinition<
+    ResponseShape,
+    ResolveQueryStrict<CoMapSchema<ResponseShape>, ResponseResolve>
+  >;
+}
+
+type AsNullablePayload<T extends MessageShape> = T extends Record<string, never>
+  ? undefined
+  : never;
+type MessageValuePayload<T extends MessageShape> =
+  | Simplify<CoMapInitZod<T>>
+  | AsNullablePayload<T>;
+
+function createMessageEnvelope<S extends MessageShape>(
+  schema: AnyCoMapSchema,
+  value: MessageValuePayload<S>,
+  owner: Account,
+  sharedWith: Account | Group,
+): Loaded<CoMapSchema<S>> {
```

Comment:
Is this return type correct? Shouldn't we be returning a CoMap instance?

Reply:
Yes, it is correct.

The CoMap instance here would be incorrect because of https://github.com/garden-co/jazz/issues/2280

Reply:
You're right, got `Loaded` mixed up with `Resolved`

Reply:
@nrainhart `Loaded<Schema>` is an instance type of that schema!